File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
  var answer = 0;
  let isPrime = makePrimeArray();
  let s = new Set();
  answer = makeNums(numbers, [0,0,0,0,0,0,0], '', isPrime, 0);

  return answer;
}

function makePrimeArray() {
  let isPrime = [];
  isPrime.length = 10000000;
  isPrime.fill(true);
  isPrime[0] = false;
  isPrime[1] = false;
  for (let i = 2; i < Math.sqrt(10000000) + 1; i++) {
    if (isPrime[i]) {
      for (let j = 2; j <= 10000000 / i; j++) {
        isPrime[i*j] = false;
      }
    }
  }
  return isPrime;
}

function makeNums(numbers, inds, num, isPrime, cnt) {
  if (cnt === numbers.length) {
    if (isPrime[+num]) {
      isPrime[+num] = false;
      return 1;
    }
    else return 0;
  }
  let answer = 0;
  if (isPrime[+num]) {
    answer++;
    isPrime[+num] = false;
  }
  for (let i = 0; i < numbers.length; i++) {
    if (inds[i] === 0) {
      inds[i] = 1;
      answer += makeNums(numbers, inds, num+numbers.at(i), isPrime, cnt+1);
      inds[i] = 0;
    }
  }
  return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from math import sqrt
from itertools import permutations

def solution(numbers):
    answer = 0
    number_list = list(numbers)
    isPrime = makePrimes()
    a = set()
    for i in range(1, len(number_list)+1):
        a |= set(map(int, map("".join, permutations(number_list, i))))
    number_set = list(a)
    for number in number_set:
        if isPrime[number]:
            answer += 1
    return answer

def makePrimes():
    l = 10000000
    isPrime = [True]*(l+1)
    isPrime[0] = False
    isPrime[1] = False
    
    for i in range(2, int(sqrt(l))+1):
        if isPrime[i]:
            for j in range(2, l//i+1):
                isPrime[i * j] = False
    return isPrime

def findPrimeNumber(number_list, used_list, isPrime, cnt, n, number):
    if cnt == n:
        number = int(number)
        if isPrime[number]:
            isPrime[number] = False
            return 1
        return 0
    count = 0
    for i in range(len(number_list)):
        if not used_list[i]:
            used_list[i] = True
            count += findPrimeNumber(number_list, used_list, isPrime, cnt+1, n, number+number_list[i])
            used_list[i] = False
            
    return count
    